tests/dcp-prototype/backend/browser/code/common/test_db_utils.py

```diff
@@ -0,0 +1,129 @@
+import unittest
+from unittest import mock
+
+from dcp_prototype.backend.browser.code.common.db_utils import DbUtils
+
+
+class TestDbUtils(unittest.TestCase):
+    def setUp(self):
+        self.project_id = "test_id"
+        self.file_id = "file_id"
+
+    @mock.patch("dcp_prototype.backend.browser.code.common.db_utils.DbUtils.query_project_species")
+    @mock.patch("dcp_prototype.backend.browser.code.common.db_utils.DbUtils.query_project_organs")
+    @mock.patch("dcp_prototype.backend.browser.code.common.db_utils.DbUtils.query_project_assays")
+    @mock.patch("dcp_prototype.backend.browser.code.common.db_utils.DbUtils._query")
+    @mock.patch("dcp_prototype.backend.browser.code.common.db_utils.DbUtils.__init__")
+    def test_query_projects(
+        self, mock_init, mock_query, mock_query_project_assays, mock_query_project_organs, mock_query_projet_species
+    ):
+        mock_init.return_value = None
+        mock_query.return_value = [mock.Mock()]
+
+        db = DbUtils()
+        projects = db.query_projects()
+        self.assertEqual(len(projects), 1)
+        self.assertIn("id", projects[0])
+        self.assertIn("title", projects[0])
+        self.assertIn("assays", projects[0])
+        self.assertIn("organs", projects[0])
+        self.assertIn("species", projects[0])
+        self.assertIn("cell_count", projects[0])
```

Comment:
So these test that the right keys are returned for the each project, but not the values? It's worth it to make sure that the right values are returned, right?

Perhaps, testing the database logic would be easier and more complete with a test database ([perhaps even in-memory](https://www.oreilly.com/library/view/essential-sqlalchemy-2nd/9781491916544/ch04.html)) complete with some fixtures. This way you wouldn't need mocks, could simply query for items as you would normally, and test that the values come through properly.

Reply:
Awesome! This was what I was looking for- an in-memory test db. I didn't find a solution so I opted for these unit tests (optimizing for sub-second run times) for sanity checks and the functional tests to verify SQL outputs. 

I've created a [ticket](https://app.zenhub.com/workspaces/dcp2-5e2a191dad828d52cc78b028/issues/chanzuckerberg/dcp-prototype/237) to improve unit tests this way. Given the simplicity of the API at the moment, the functional tests will cover E2E functionality including SQL.